### PR TITLE
JS: Add query for unsafe construction of code from library input

### PR DESCRIPTION
**Pull request by erik-krogh**
CVE-2017-5954: TP/FP (conditional unsafety)
CVE-2019-5413: TP/TN
CVE-2020-28502: TP/TN
CVE-2020-7712: TP/TN
CVE-2021-23406: TP/TN
CVE-2021-23337: TP
CVE-2021-32831: TP
CVE-2021-23358: TP
CVE-2020-7640: TP/TN

[A test run](https://lgtm.com/query/6107640250463390809/) shows a lot of results, many of which just looks to be (very) bad style.  
So I've set the severity to `warning`. 

[An evaluation looks good.](https://github.com/github/codeql-dca-main/tree/run/erik-krogh/libCode-eval/reports). (The baseline is `main` + a `where none()` edition of the query).   
The new results look reasonable to me, even though they're basically just bad style.  
(I suppose there are performance reasons for implementing it in that style).

### CODE REVIEW

File: javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+Dynamically constructing code with inputs from exported functions 
+may allow compromise clients that use these functions with untrusted inputs.
```
Comment (intrigus-lgtm) [inline javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp:9]: This sentence looks incomplete.
"may allow compromise clients" ??

File: javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+Dynamically constructing code with inputs from exported functions
```
Comment (esbena) [inline javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp:8]: Same formulation concerns as in #5769

File: javascript/ql/src/semmle/javascript/security/dataflow/UnsafeCodeConstruction.qll [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    // override to require that there is a path without unmatched return steps
+    override predicate hasFlowPath(DataFlow::SourcePathNode source, DataFlow::SinkPathNode sink) {
+      super.hasFlowPath(source, sink) and
+      exists(DataFlow::MidPathNode mid |
+        source.getASuccessor*() = mid and
+        sink = mid.getASuccessor() and
+        mid.getPathSummary().hasReturn() = false
+      )
+    }
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/UnsafeCodeConstruction.qll:49]: Rebase on top of #5769 , and reuse the similar predicate from that PR

File: javascript/ql/src/semmle/javascript/security/dataflow/UnsafeCodeConstructionCustomizations.qll [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A parameter of an exported function, seen as a source.
+   */
+  class ExternalInputSource extends Source, DataFlow::ParameterNode {
+    ExternalInputSource() {
+      this = Exports::getALibraryInputParameter() and
+      not this.getName() = "code"
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/UnsafeCodeConstructionCustomizations.qll:27]:
```suggestion
      // permit parameters that clearly are intended to contain executable code.
      not this.getName() = "code"
```

Comment (esbena) [reply javascript/ql/src/semmle/javascript/security/dataflow/UnsafeCodeConstructionCustomizations.qll:27]: (go home diff, you are drunk :) )

**Review by esbena — COMMENTED** [review]
LGTM, but lets merge #5769 first, and then apply the teachings from that PR here.

**Comment by asgerf** [comment]
Sorry for leaving this PR hanging for so long. Looking at the evaluation results, it looks like all of them are FPs:
- The Angular results are in internal API. I've put up a PR to fix it here: https://github.com/github/codeql/pull/6789
- The result in Underscore is a template compiler, which by design can evaluate JS code. It's not bad style or a performance trick; these templates can contain JS code which should be evaluated as-is.

I'd say templating engines should be considered a well-known counter example, and an allow-list should be used to exclude methods with names like `template` or `compile` or perhaps even `render`.

**Comment by esbena** [comment]
> I'd say templating engines should be considered a well-known counter example, and an allow-list should be used to exclude methods with names like template or compile or perhaps even render.

I agree with this. Could we use a whitelist to avoid intentional APIs like those and otherwise merge this twice-stalled PR as is? We can then separately address the undesirable FNs the whitelist causes.

**Comment by erik-krogh** [comment]
> I agree with this. Could we use a whitelist to avoid intentional APIs like those and otherwise merge this twice-stalled PR as is? We can then separately address the undesirable FNs the whitelist causes.

I tried to look into the method-names of the CVEs currently flagged by this query: 

| CVE | method-name|
| --- | --- | 
| CVE-2017-5954 | deserialize |
| CVE-2019-5413 | compile |
| CVE-2020-28502 | send |
| CVE-2020-7712 | parseLookup |
| CVE-2021-23406 | compile |
| CVE-2021-23337 | template |
| CVE-2021-32831 | set |
| CVE-2021-23358 | template |
| CVE-2020-7640 | create |

Adding a allow-list based on names that are template-engine-like would remove some of those TPs.  
I would prefer we lower the precision of the query without an allow-list.

File: javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.ql [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/**
+ * @name Unsafe code constructed from libary input
+ * @description Using externally controlled strings to construct code may allow a malicious
+ *              user to execute arbitrary code.
+ * @kind path-problem
+ * @problem.severity warning
+ * @precision warning
```
Comment (esbena) [inline javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.ql:7]:
```suggestion
 * @precision medium
```
:)

Comment (esbena) [reply javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.ql:7]: We should have a ql-for-ql query for this...

File: javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+<p>
+Properly document library functions that construct code from unsanitized 
+inputs, or avoid constructing code in the first place.
+</p>
+</recommendation>
+
+<example>
+<p>
+The following example shows two methods implemented using `eval`: a simple
+deserialization routine and a getter method.
```
Comment (ethanpalm) [inline javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp:29]:
```suggestion
The following example shows two methods implemented using `eval`: a simple
deserialization routine and a getter method.
If untrusted inputs are used with these methods,
then an attacker might be able to execute arbitrary code on the system. 
```

File: javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+The following example shows two methods implemented using `eval`: a simple
+deserialization routine and a getter method. 
+</p>
+
+<sample src="examples/UnsafeCodeConstruction.js" />
+
+<p>
+If untrusted inputs are used with these methods,
+then an attacker might be able to execute arbitrary code on the system.
+</p>
```
Comment (ethanpalm) [inline javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp:37]:
```suggestion
```

File: javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp (hunk 6 above)
Comment (ethanpalm) [reply javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp:29]: Let's move this above the example to explain why the example is a bad practice before people read it.

File: javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+<p>
+To avoid this problem, either properly document that the function is potentially
+unsafe, or use an alternative solution such as `JSON.parse` or another library
```
Comment (ethanpalm) [inline javascript/ql/src/Security/CWE-094/UnsafeCodeConstruction.qhelp:40]:
```suggestion
unsafe, or use an alternative solution such as `JSON.parse` or another library, like in the examples below, 
```

**Review by ethanpalm — DISMISSED** [review]
This looks good from a docs point of view. I am just making one small change to move the explanation of why the first example is a bad coding practice to the paragraph before the example rather than after the example ⚡

**Review by ethanpalm — APPROVED** [review]

**Pull request merged**